🐛Cache issue with identical files with relative import

It seems like there is a cache issue when you have two identical files in two different folder, but they have relative import path. As an example, I have `folder1/subFolder/index.js` and `folder2/subFolder/index.js` importing `../index.js`. However, `folder1/index.js` and `folder2/index.js` are two different files. 

It looks like Parcel is thinking they are both the same but they are not because of the relative path.

### 😯 Current Behavior

Here is a proof of concept: https://github.com/exon/parcel-cache-bug

If you build and run it, it will print:
```
FOLDER1!!!
FOLDER3!!!
```

### 🤔 Expected Behavior

However, the expected result were:
```
FOLDER1!!!
FOLDER2!!!
FOLDER3!!!
```

### 💁 Possible Solution

I don't know the Parcel cache system but I guess it should transform the import path into something more absolute starting from the project root.

@eXon I reproduced your issue (I did have to tweak the `start` script in your `package.json` from `"start": "node dist/parcel-cache-bug"` to `"start": "node dist/index.js"`).

I ruled out caching being the culprit as the behavior still occurs even after I nuked the cache (`rm -rf .cache`) and ran `yarn build --no-cache` to effectively run `parcel build index.js --no-cache`.

I de-minified `dist/index.js`:

```javascript
require = function(r, e, n) {
    var t = "function" == typeof require && require;

    function o(n, u) {
        if (!e[n]) {
            if (!r[n]) {
                var i = "function" == typeof require && require;
                if (!u && i) return i(n, !0);
                if (t) return t(n, !0);
                var f = new Error("Cannot find module '" + n + "'");
                throw f.code = "MODULE_NOT_FOUND", f
            }
            l.resolve = function(e) {
                return r[n][1][e] || e
            };
            var c = e[n] = new o.Module(n);
            r[n][0].call(c.exports, l, c, c.exports)
        }
        return e[n].exports;

        function l(r) {
            return o(l.resolve(r))
        }
    }
    o.isParcelRequire = !0, o.Module = function(r) {
        this.id = r, this.bundle = o, this.exports = {}
    }, o.modules = r, o.cache = e, o.parent = t;
    for (var u = 0; u < n.length; u++) o(n[u]);
    return o
}({
    9: [function(require, module, exports) {
        module.exports = function() {
            console.log("FOLDER1!!!")
        };
    }, {}],
    3: [function(require, module, exports) {
        var r = require("../");
        r();
    }, {
        "../": 9
    }],
    11: [function(require, module, exports) {
        module.exports = function() {
            console.log("FOLDER2!!!")
        };
    }, {}],
    13: [function(require, module, exports) {
        module.exports = function() {
            console.log("FOLDER3!!!")
        };
    }, {}],
    7: [function(require, module, exports) {
        var global = (1, eval)("this");
        var e = (0, eval)("this"),
            r = require("../");
        r(), e.different = !0;
    }, {
        "../": 13
    }],
    1: [function(require, module, exports) {
        require("./folder1/subfolder"), require("./folder2/subfolder"), require("./folder3/subfolder");
    }, {
        "./folder1/subfolder": 3,
        "./folder2/subfolder": 3,
        "./folder3/subfolder": 7
    }]
}, {}, [1])
//# sourceMappingURL=/dist/index.map
```

Looks like at the bottom both `folder1` and `folder2` are being set to `3` presumably because the actual text they contain is the same (note that even if you had used different variable names across `folder2/subfolder/index.js` or `folder3/subfolder/index.js`, parcel ends up changing those and abstracting them so its not sufficient just to have different variable names, the actual *values* in the code have to be different).

I'm going to do some more digging to see what part of Parcel could be causing the issue...

--- a/src/packagers/JSPackager.js
+++ b/src/packagers/JSPackager.js
@@ -35,13 +35,13 @@ class JSPackager extends Packager {
   }
 
   async addAsset(asset) {
-    if (this.dedupe.has(asset.generated.js)) {
+    if (this.dedupe.has(this.dedupeKey(asset))) {
       return;
     }
 
     // Don't dedupe when HMR is turned on since it messes with the asset ids
     if (!this.options.hmr) {
-      this.dedupe.set(asset.generated.js, asset.id);
+      this.dedupe.set(this.dedupeKey(asset), asset.id);
     }
 
     let deps = {};
@@ -60,7 +60,7 @@ class JSPackager extends Packager {
         deps[dep.name] = bundles;
         this.bundleLoaders.add(mod.type);
       } else {
-        deps[dep.name] = this.dedupe.get(mod.generated.js) || mod.id;
+        deps[dep.name] = this.dedupe.get(this.dedupeKey(mod)) || mod.id;
 
         // If the dep isn't in this bundle, add it to the list of external modules to preload.
         // Only do this if this is the root JS bundle, otherwise they will have already been
@@ -91,6 +91,17 @@ class JSPackager extends Packager {
     }
 
     return name;
+  }
+
+  dedupeKey(asset) {
+    // cannot rely *only* on generated JS for deduplication because paths like
+    // `../` can cause 2 identical JS files to behave differently depending on
+    // where they are located on the filesystem
+    let deps = Array.from(asset.depAssets.keys(), dep =>
+      this.bundler.resolver.resolveFilename(dep.name, asset.name)
+    );
+    deps.sort();
+    return JSON.stringify([asset.generated.js, deps]);
   }
 
   async writeModule(id, code, deps = {}, map) {

--- a/src/packagers/JSPackager.js
+++ b/src/packagers/JSPackager.js
@@ -3,6 +3,7 @@ const path = require('path');
 const Packager = require('./Packager');
 const urlJoin = require('../utils/urlJoin');
 const lineCounter = require('../utils/lineCounter');
+const objectHash = require('../utils/objectHash');
 
 const prelude = {
   source: fs
@@ -35,13 +36,14 @@ class JSPackager extends Packager {
   }
 
   async addAsset(asset) {
-    if (this.dedupe.has(this.dedupeKey(asset))) {
+    let key = this.dedupeKey(asset);
+    if (this.dedupe.has(key)) {
       return;
     }
 
     // Don't dedupe when HMR is turned on since it messes with the asset ids
     if (!this.options.hmr) {
-      this.dedupe.set(this.dedupeKey(asset), asset.id);
+      this.dedupe.set(key, asset.id);
     }
 
     let deps = {};
@@ -97,11 +99,8 @@ class JSPackager extends Packager {
     // cannot rely *only* on generated JS for deduplication because paths like
     // `../` can cause 2 identical JS files to behave differently depending on
     // where they are located on the filesystem
-    let deps = Array.from(asset.depAssets.keys(), dep =>
-      this.bundler.resolver.resolveFilename(dep.name, asset.name)
-    );
-    deps.sort();
-    return JSON.stringify([asset.generated.js, deps]);
+    let deps = Array.from(asset.depAssets.values(), dep => dep.name).sort();
+    return objectHash([asset.generated.js, deps]);
   }
 
   async writeModule(id, code, deps = {}, map) {

--- a/test/integration/js-different-contents/hello.js
+++ b/test/integration/js-different-contents/hello.js
@@ -1,0 +1,2 @@
+const value = 'Hello'
+export default value

--- a/test/integration/js-different-contents/index.js
+++ b/test/integration/js-different-contents/index.js
@@ -1,0 +1,4 @@
+import hello from './hello.js'
+import world from './world.js'
+
+export default `${hello} ${world}!`

--- a/test/integration/js-different-contents/world.js
+++ b/test/integration/js-different-contents/world.js
@@ -1,0 +1,2 @@
+const value = 'World'
+export default value

--- a/test/integration/js-same-contents-different-dependencies/hello/index.js
+++ b/test/integration/js-same-contents-different-dependencies/hello/index.js
@@ -1,0 +1,3 @@
+// hello/index.js and world/index.js are exactly the same content-wise, but in different locations
+import value from './value'
+export default value

--- a/test/integration/js-same-contents-different-dependencies/hello/value.js
+++ b/test/integration/js-same-contents-different-dependencies/hello/value.js
@@ -1,0 +1,2 @@
+const value = 'Hello'
+export default value

--- a/test/integration/js-same-contents-different-dependencies/index.js
+++ b/test/integration/js-same-contents-different-dependencies/index.js
@@ -1,0 +1,4 @@
+import hello from './hello'
+import world from './world'
+
+export default `${hello} ${world}!`

--- a/test/integration/js-same-contents-different-dependencies/world/index.js
+++ b/test/integration/js-same-contents-different-dependencies/world/index.js
@@ -1,0 +1,3 @@
+// hello/index.js and world/index.js are exactly the same content-wise, but in different locations
+import value from './value'
+export default value

--- a/test/integration/js-same-contents-different-dependencies/world/value.js
+++ b/test/integration/js-same-contents-different-dependencies/world/value.js
@@ -1,0 +1,2 @@
+const value = 'World'
+export default value

--- a/test/integration/js-same-contents-same-dependencies/hello1.js
+++ b/test/integration/js-same-contents-same-dependencies/hello1.js
@@ -1,0 +1,2 @@
+const value = 'Hello'
+export default value

--- a/test/integration/js-same-contents-same-dependencies/hello2.js
+++ b/test/integration/js-same-contents-same-dependencies/hello2.js
@@ -1,0 +1,2 @@
+const value = 'Hello'
+export default value

--- a/test/integration/js-same-contents-same-dependencies/index.js
+++ b/test/integration/js-same-contents-same-dependencies/index.js
@@ -1,0 +1,4 @@
+import hello1 from './hello1'
+import hello2 from './hello2'
+
+export default `${hello1} ${hello2}!`

--- a/test/javascript.js
+++ b/test/javascript.js
@@ -977,7 +977,7 @@ describe('javascript', function() {
       }
     );
 
-    let module = run(b);
+    let module = await run(b);
     assert.equal(module.default, 'Hello World!');
   });
 
@@ -990,7 +990,7 @@ describe('javascript', function() {
       }
     );
 
-    let module = run(b);
+    let module = await run(b);
     assert.equal(module.default, 'Hello World!');
   });
 
@@ -1016,7 +1016,7 @@ describe('javascript', function() {
       )
     );
 
-    let module = run(b);
+    let module = await run(b);
     assert.equal(module.default, 'Hello Hello!');
   });
 });

--- a/test/javascript.js
+++ b/test/javascript.js
@@ -1004,15 +1004,15 @@ describe('javascript', function() {
     const {rootDir} = b.entryAsset.options;
     const dedupedAssets = Array.from(b.offsets.keys()).map(asset => asset.name);
     assert.equal(dedupedAssets.length, 2);
-    assert(dedupedAssets.includes(`${rootDir}/index.js`));
+    assert(dedupedAssets.includes(path.join(rootDir, 'index.js')));
     assert(
-      dedupedAssets.includes(`${rootDir}/hello1.js`) ||
-        dedupedAssets.includes(`${rootDir}/hello2.js`)
+      dedupedAssets.includes(path.join(rootDir, 'hello1.js')) ||
+        dedupedAssets.includes(path.join(rootDir, 'hello2.js'))
     );
     assert(
       !(
-        dedupedAssets.includes(`${rootDir}/hello1.js`) &&
-        dedupedAssets.includes(`${rootDir}/hello2.js`)
+        dedupedAssets.includes(path.join(rootDir, 'hello1.js')) &&
+        dedupedAssets.includes(path.join(rootDir, 'hello2.js'))
       )
     );
 

--- a/test/javascript.js
+++ b/test/javascript.js
@@ -968,4 +968,55 @@ describe('javascript', function() {
     await run(b, {define: mockDefine, module: undefined});
     assert.equal(test, 2);
   });
+
+  it('should not dedupe imports with different contents', async function() {
+    let b = await bundle(
+      __dirname + `/integration/js-different-contents/index.js`,
+      {
+        hmr: false // enable asset dedupe in JSPackager
+      }
+    );
+
+    let module = run(b);
+    assert.equal(module.default, 'Hello World!');
+  });
+
+  it('should not dedupe imports with same content but different absolute dependency paths', async function() {
+    let b = await bundle(
+      __dirname +
+        `/integration/js-same-contents-different-dependencies/index.js`,
+      {
+        hmr: false // enable asset dedupe in JSPackager
+      }
+    );
+
+    let module = run(b);
+    assert.equal(module.default, 'Hello World!');
+  });
+
+  it('should dedupe imports with same content and same dependency paths', async function() {
+    let b = await bundle(
+      __dirname + `/integration/js-same-contents-same-dependencies/index.js`,
+      {
+        hmr: false // enable asset dedupe in JSPackager
+      }
+    );
+    const {rootDir} = b.entryAsset.options;
+    const dedupedAssets = Array.from(b.offsets.keys()).map(asset => asset.name);
+    assert.equal(dedupedAssets.length, 2);
+    assert(dedupedAssets.includes(`${rootDir}/index.js`));
+    assert(
+      dedupedAssets.includes(`${rootDir}/hello1.js`) ||
+        dedupedAssets.includes(`${rootDir}/hello2.js`)
+    );
+    assert(
+      !(
+        dedupedAssets.includes(`${rootDir}/hello1.js`) &&
+        dedupedAssets.includes(`${rootDir}/hello2.js`)
+      )
+    );
+
+    let module = run(b);
+    assert.equal(module.default, 'Hello Hello!');
+  });
 });

